### PR TITLE
解决issue#6630问题：heartbeat配置bug

### LINKED ISSUE #6630: heeratbeat配置bug，导致连接异常关闭
- version：2.7.8

image

dubbo现在的机制是如果consumer没有配置心跳时间，但是provider配置了心跳则在consmer合并url的时候以provider为准。

默认使用Netty4的idle-handler进行空闲检测，

## 场景一：不开启简化注册中心url，且consumer配置心跳间隔大于provider的3倍，且`send.reconect`属性为缺省值fasle

ServiceB（provider）配置heartbeat时间为2s，此时ServiceB将在2s*3=6s后断开长连接。

ServiceA（consumer）配置heartbeat时间为30s，此时ServiceA将在30s后发送第一次跳。

所以在实际上此时连接已经断开，并且根据idle-handler的机制，当连接关闭后已经不会再发送跳，并且在调用时不会重新建立连接，导致异常。

## 场景二：启简化注册中心url，配置porvider的heartbeat，consumer使用缺省值，且`send.reconect`属性为缺省值fasle

ServiceB（provider）配置heartbeat时间为2s，此时ServiceB将在2s*3=6s后断开长连接。

ServiceA（consumer）使用缺省heartbeat值60s，按consumer合并url策略，应该使用2s为心跳间隔，但是简化注册中心后heartbeat属性被忽略掉了。

导致场景一异常再次出现。

## 改进方案

1. 简化注册中心url，加上heartbeat。
2. consumer合并url时和provider的heartbeat进行比较，如果consumer.heartbeat >= provider.heartbeat * 3。则需要以provider.heartbeat为准，保证不出现连接关闭的情况。
3. 目前consumer不支持全局配置heartbeat和protocol，不能全局配置心跳间隔，应该支持使用protocol全局配置。

### PR DESCRIPTION
**Pull request by wangxw666**
fix #6630

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/config/ReferenceConfigBase.java [hunk 1]
```diff
@@ -67,6 +68,11 @@
      */
     protected ConsumerConfig consumer;
 
+    /**
+     * The protocol config
+     */
+    protected ProtocolConfig _protocol;
```
[the remark this answers is not in the document]
Comment (chickenlj) [reply dubbo-common/src/main/java/org/apache/dubbo/config/ReferenceConfigBase.java:74]: Yes, I agree better to follow the name convention as the other existing properties.

**Comment by chickenlj** [comment]
please fix the failed CI.

[INFO] There is 1 error reported by Checkstyle 8.9 with codestyle/checkstyle.xml ruleset.
[ERROR] src/main/java/org/apache/dubbo/registry/integration/RegistryProtocol.java:[99] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - org.apache.dubbo.remoting.Constants.*.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:

File: dubbo-common/src/main/java/org/apache/dubbo/config/ReferenceConfigBase.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (IndiraFinish) [reply dubbo-common/src/main/java/org/apache/dubbo/config/ReferenceConfigBase.java]: ok

**Comment by AlbumenJ** [comment]
@XiaoweiM hi, thanks for your contribution.

Please resolve conficts with the latest master branch.

**Comment by XiaoweiM** [comment]
ok

**Comment by AlbumenJ** [comment]
Close for long time no response.

Please feel free to reopen if you have any question.

If you think these changes still useful in the latest master branch, please submit a new pull request.

**Pull request closed without merging**